katdal/datasources.py

```diff
@@ -102,6 +103,80 @@ def _apply_data_lost(orig_flags, lost, block_id):
     return flags
 
 
+def corrprod_to_autocorr(corrprod):
+    """Find the autocorrelation indices of correlation products.
+
+    Parameters
+    ----------
+    corrprod : sequence of 2-tuples or ndarray
+        Input labels of the correlation products
+
+    Returns
+    -------
+    auto_indices : np.ndarray
+        The indices in corrprod that correspond to auto-correlations
+    index1, index2 : np.ndarray
+        Lists of the same length as corrprod, containing the indices within
+        `auto_indices` referring to the first and second corresponding
+        autocorrelations.
+
+    Raises
+    ------
+    KeyError
+        If any of the autocorrelations are missing
+    """
+    auto_indices = []
+    auto_lookup = {}
+    for i, baseline in enumerate(corrprod):
+        if baseline[0] == baseline[1]:
+            auto_lookup[baseline[0]] = len(auto_indices)
+            auto_indices.append(i)
+    index1 = [auto_lookup[a] for (a, b) in corrprod]
+    index2 = [auto_lookup[b] for (a, b) in corrprod]
+    return np.array(auto_indices), np.array(index1), np.array(index2)
+
+
+@numba.jit(nopython=True, nogil=True)
+def weight_power_scale(block, auto_indices, index1, index2, out=None, tmp=None):
+    """Compute weight scale from visibility data.
+
+    This function is designed to be usable with
+    :func:`dask.array.map_blocks`.
+
+    Parameters
+    ----------
+    block : np.ndarray
+        Chunk of visibility data, with dimensions time, frequency, baseline
+        (or any two dimensions then baseline). It must contain all the
+        baselines of a stream.
+    auto_indices, index1, index2 : np.ndarray
+        Arrays returned by :func:`corrprod_to_autocrr`
+    out : np.ndarray, optional
+        If specified, the output array, with same shape as `block` and dtype ``np.float32``
+    tmp : np.ndarray, optional
+        If specified, an array to use as internal scratch space (useful to
+        reuse the space across multiple calls). It must have the same shape
+        as `auto_indices` and dtype ``np.float32``.
+    """
+    auto_scale = np.empty(len(auto_indices), np.float32) if tmp is None else tmp
+    power_scale = np.empty(block.shape, np.float32) if out is None else out
+    bad_weight = np.float32(2.0**-32)
+    for i in range(block.shape[0]):
+        for j in range(block.shape[1]):
+            for k in range(len(auto_indices)):
+                auto_scale[k] = np.divide(1, block[i, j, auto_indices[k]].real)
```

Comment:
What is the difference between this and `1 / block...`?

Reply:
Good question. The latter raises ZeroDivisionError (might be a numba quirk), but I want to deal with that lower down (np.isfinite check). I've added a comment to explain it.

Reply:
I get similar behaviour to `np.divide` with straight NumPy arrays, e.g.

```Python
1 / np.array(0)
```

The ZeroDivisionError only appears when NumPy is not involved:

```
1 / 0
```